# of replicas for net-contour-controller not updated by v1alpha1.HighAvailability.Replicas

**Describe the bug**
Got a reconciliation timeout on the `test/e2e: TestKnativeServingDeployment` with a `contour` network after the update #749

```
    verify.go:301: Retrying...
    verify.go:300: deployment "net-contour-controller": Status.Replicas = 1, want: 2
    verify.go:301: Retrying...
    verify.go:300: deployment "net-contour-controller": Status.Replicas = 1, want: 2
    verify.go:301: Retrying...
    verify.go:308: Timed out waiting on KnativeServing to delete timed out waiting for the condition
    --- FAIL: TestKnativeServingDeployment/configure (354.50s)
```

```
$ kubectl get deploy -n knative-operator-testing
NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
activator                2/2     2            2           3m53s
autoscaler               2/2     2            2           3m53s
autoscaler-hpa           2/2     2            2           3m50s
controller               2/2     2            2           3m53s
domain-mapping           2/2     2            2           3m52s
domainmapping-webhook    2/2     2            2           3m52s
net-contour-controller   1/1     1            1           3m50s
net-istio-controller     2/2     2            2           2m47s
net-istio-webhook        2/2     2            2           2m47s
webhook                  2/2     2            2           3m51s

```

Full log: https://storage.googleapis.com/knative-prow/logs/ci-knative-operator-s390x-e2e-tests/1442776110814203904/build-log.txt

**Expected behavior**
\# of replicas for `net-contour-controller` is also updated to 2 like other components under `KnativeServing`

**To Reproduce**
From a cleanly-installed k8s cluster,

```
$ export KO_DOCKER_REPO=localhost:5000
$ export INGRESS_CLASS=contour.ingress.networking.knative.dev
$ export KUBECONFIG=$HOME/.kube/config
$ kubectl create ns knative-serving
$ # istio only needed for `gateways.networking.istio.io` CRD 
$ kubectl apply -f https://github.com/knative-sandbox/net-istio/releases/download/v0.23.0/istio.yaml 2> /dev/null
$ kubectl apply -f https://raw.githubusercontent.com/knative/serving/main/third_party/contour-latest/contour.yaml
$ kubectl apply -f https://raw.githubusercontent.com/knative/serving/main/third_party/contour-latest/net-contour.yaml
$ kubectl wait --timeout=5m pod --for=condition=Ready -n contour-external -l '!job-name'
$ ./test/e2e-tests.sh --run-tests
```

**Knative release version**
main branch (1403d9c2dc27b32a2bff8e5e500acd0a05661779)

--- a/test/resources/knativeserving.go
+++ b/test/resources/knativeserving.go
@@ -60,19 +60,6 @@ func WaitForKnativeServingState(clients servingv1alpha1.KnativeServingInterface,
 
 // EnsureKnativeServingExists creates a KnativeServing with the name names.KnativeServing under the namespace names.Namespace, if it does not exist.
 func EnsureKnativeServingExists(clients servingv1alpha1.KnativeServingInterface, names test.ResourceNames) (*v1alpha1.KnativeServing, error) {
-	ingressClass := "istio.ingress.networking.knative.dev"
-	if ingressClassOverride := os.Getenv("INGRESS_CLASS"); ingressClassOverride != "" {
-		ingressClass = ingressClassOverride
-	}
-	var istioEnabled, contourEnabled, kourierEnabled bool
-	switch strings.Split(ingressClass, ".")[0] {
-	case "istio":
-		istioEnabled = true
-	case "contour":
-		contourEnabled = true
-	case "kourier":
-		kourierEnabled = true
-	}
 	// If this function is called by the upgrade tests, we only create the custom resource, if it does not exist.
 	ks, err := clients.Get(context.TODO(), names.KnativeServing, metav1.GetOptions{})
 	if apierrs.IsNotFound(err) {
@@ -81,17 +68,8 @@ func EnsureKnativeServingExists(clients servingv1alpha1.KnativeServingInterface,
 				Name:      names.KnativeServing,
 				Namespace: names.Namespace,
 			},
-			Spec: v1alpha1.KnativeServingSpec{
-				Ingress: &v1alpha1.IngressConfigs{
-					Istio:   v1alpha1.IstioIngressConfiguration{Enabled: istioEnabled},
-					Contour: v1alpha1.ContourIngressConfiguration{Enabled: contourEnabled},
-					Kourier: v1alpha1.KourierIngressConfiguration{Enabled: kourierEnabled},
-				},
-				CommonSpec: v1alpha1.CommonSpec{
-					Config: v1alpha1.ConfigMapData{"network": {"ingress.class": ingressClass}},
-				},
-			},
 		}
+		configureIngressClass(&ks.Spec)
 		return clients.Create(context.TODO(), ks, metav1.CreateOptions{})
 	}
 	return ks, err
@@ -129,7 +107,7 @@ func getDeploymentStatus(d *v1.Deployment) corev1.ConditionStatus {
 }
 
 func getTestKSOperatorCRSpec() v1alpha1.KnativeServingSpec {
-	return v1alpha1.KnativeServingSpec{
+	spec := v1alpha1.KnativeServingSpec{
 		CommonSpec: v1alpha1.CommonSpec{
 			Config: v1alpha1.ConfigMapData{
 				DefaultsConfigKey: {
@@ -141,5 +119,37 @@ func getTestKSOperatorCRSpec() v1alpha1.KnativeServingSpec {
 				},
 			},
 		},
+	}
+	configureIngressClass(&spec)
+	return spec
+}
+
+func configureIngressClass(spec *v1alpha1.KnativeServingSpec) {
+	ingressClass := "istio.ingress.networking.knative.dev"
+	if ingressClassOverride := os.Getenv("INGRESS_CLASS"); ingressClassOverride != "" {
+		ingressClass = ingressClassOverride
+	}
+	var istioEnabled, contourEnabled, kourierEnabled bool
+	switch strings.Split(ingressClass, ".")[0] {
+	case "istio":
+		istioEnabled = true
+	case "contour":
+		contourEnabled = true
+	case "kourier":
+		kourierEnabled = true
+	}
+
+	if !istioEnabled {
+		spec.Ingress = &v1alpha1.IngressConfigs{
+			Istio:   v1alpha1.IstioIngressConfiguration{Enabled: istioEnabled},
+			Contour: v1alpha1.ContourIngressConfiguration{Enabled: contourEnabled},
+			Kourier: v1alpha1.KourierIngressConfiguration{Enabled: kourierEnabled},
+		}
+
+		if spec.CommonSpec.Config == nil {
+			spec.CommonSpec.Config = v1alpha1.ConfigMapData{"network": {"ingress.class": ingressClass}}
+		} else {
+			spec.CommonSpec.Config["network"] = map[string]string{"ingress.class": ingressClass}
+		}
 	}
 }

--- a/test/resources/verify.go
+++ b/test/resources/verify.go
@@ -65,14 +65,14 @@ func KSOperatorCRVerifyConfiguration(t *testing.T, clients *test.Clients, names 
 	// Delete a single key/value pair
 	verifySingleKeyDeletion(t, LoggingConfigKey, loggingConfigMapName, clients, names)
 
+	// Verify HA config
+	VerifyHADeployments(t, clients, names)
+
 	// Use an empty map as the value
 	verifyEmptyKey(t, DefaultsConfigKey, defaultsConfigMapName, clients, names)
 
 	// Now remove the config from the spec and update
 	verifyEmptySpec(t, loggingConfigMapName, clients, names)
-
-	// Verify HA config
-	VerifyHADeployments(t, clients, names)
 }
 
 func verifyDefaultConfig(t *testing.T, ks *v1alpha1.KnativeServing, defaultsConfigMapName string, clients *test.Clients, names test.ResourceNames) {

